volumes/repository.go

```diff
@@ -152,42 +109,122 @@ func (r *Repository) Delete(path string) error {
 		return err
 	}
 
-	if !volume.IsBindMount {
-		if err := r.driver.Remove(volume.ID); err != nil {
-			if !os.IsNotExist(err) {
-				return err
-			}
-		}
+	if err := volume.remove(); err != nil {
+		return err
 	}
 
 	delete(r.volumes, volume.Path)
+	delete(r.idIndex, volume.ID)
 	return nil
 }
 
-func (r *Repository) createNewVolumePath(id string) (string, error) {
-	if err := r.driver.Create(id, ""); err != nil {
-		return "", err
+func (r *Repository) FindOrCreateVolume(path, driver string) (*Volume, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if v := r.get(path); v != nil {
+		logrus.Debugf("found existing volume for: %s %v", driver)
+		if !v.exists() {
+			// Recreates the underlying volume: #10146
+			logrus.Debugf("recreating volume %s", v.Path)
+			if err := v.create(); err != nil {
+				return nil, err
+			}
+		}
+		return v, nil
+	}
+
+	v, err := r.newVolume(path, driver)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := v.toDisk(); err != nil {
+		v.remove()
+		return nil, err
+	}
+	r.add(v)
+	return v, nil
+}
+
+func (r *Repository) newVolume(path, driver string) (*Volume, error) {
+	id, err := r.generateId()
+	if err != nil {
+		return nil, err
+	}
+
+	if path == "" && driver == "" {
+		driver = "vfs"
+	}
+
+	if driver == "" {
+		driver = r.getDriverNameFromPath(path)
+		if driver == "" {
```

Comment:
Even if it doesn't return an empty string, I still prefer to check because it is technically possible as the only contract is that the function returns a string and a string can be empty.